source/melpomene/src/main.rs

```diff
@@ -347,13 +348,29 @@ fn kernel_entry(opts: MelpomeneOptions) {
         }
     });
 
+    let mut t0 = tokio::time::Instant::now();
     loop {
         while !KERNEL_LOCK.load(Ordering::Acquire) {
             sleep(Duration::from_millis(10));
         }
 
-        k.tick();
+        let tick = k.tick();
 
         KERNEL_LOCK.store(false, Ordering::Release);
+
+        // advance the timer
+        let ticks = t0.elapsed().as_millis() as u64;
+        let turn = k.timer().force_advance_ticks(ticks);
+        t0 = tokio::time::Instant::now();
+
+        if turn.expired == 0 && !tick.has_remaining {
```

Comment:
This seems generally acceptable to me, I'm not sure how we'd abstract that across other hardware platforms, but its okay if you don't solve "why are all these OS bits in the simulator and not the kernel" in this PR.

Reply:
yeah...the fact that this needs to actually figure out how much time has elapsed makes this a little more complex than just taking a "wait for interrupt" function or whatever. we'll need some kind of abstraction which i was thinking was better to figure out later.